### ISSUE TITLE
Sparse tensor values()/indices() return read-only (const) tensors

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#1438]
**tl;dr** I propose introducing a new read-only view on tensors, which disable in place operations on a tensor (but can otherwise share storage with a mutable view). `values()` and `indices()` should return these read-only views, which make it easier for sparse tensor to maintain its internal invariants.

**Motivation.** The current sparse tensors API provides two methods: `values()` and `indices()`, which let you see into the underlying representation of a sparse tensor. These are full, honest to god tensors, supporting the full complement of inplace and out-of-place operations. However, the inplace operations allow for plenty of mischief:

```
>> x = torch.sparse.FloatTensor(i,v)
>>> x
FloatTensor with indices:

 3  5
[torch.LongTensor of size 1x2]
and values:

 0
 2
[torch.FloatTensor of size 2]

>>> x.is_coalesced()
False

>>> y = x.coalesce()
>>> y.indices().fill_(0)

 0  0
[torch.LongTensor of size 1x2]

>>> y.is_coalesced()
True
>>> y
FloatTensor with indices:

 0  0
[torch.LongTensor of size 1x2]
and values:

 0
 2
[torch.FloatTensor of size 2]
```

This is a failure of abstraction: the ability to perform in-place updates on the internal tensors of a sparse tensor means that a client can break invariants that we want to apply to a sparse tensor. Unfortunately, being able to directly handle values()/indices() is quite useful, because it allows users to implement operations on sparse tensors that are not otherwise available (see #1369 for more context.)

The way to solve this situation is to outlaw inplace operations on sparse tensor internals. To do this, we need a way of providing a "read-only" view on tensor.

**Design.** I propose the following changes:

1. Add a new field to `THTensor` tracking whether or not this tensor's view on the underlying storage is read-only or not. Since there is already a `char flag` field, an extra `char` of storage should not cause the size of a padded THTensor to increase.

2. For all inplace operations, add a test if THTensor is read-only. If it is, raise an error.

3. Modify `values` and `indices` to mark the newly allocated THTensor as read-only.

4. Add new method `const` to tensor which gives you a new THTensor that has a read-only view on the underlying storage, as well as an `unsafe_mutable` to get a writable view.

**Alternatives.** Arguably, only `indices()` needs to be read-only to preserve invariants on sparse tensors; although it is tricky to handle uncoalesced `values()` correctly, in-place updates cannot cause problems. (Notably, because the sparse tensor always knows its dimensions, sparse tensors handle shape changes in their underlying tensors correctly.)

**Going further.** These are only read-only views, because it is possible for the underlying storage to change under you due to an inplace update from a writeable view on the same storage. It may also be interesting to support truly immutable storage, for which ALL views are read-only. I don't have a good use case for something like this, but maybe someone else does.

**Comment by ezyang** [comment]
@soumith and @colesbury vetoed this proposal, on the grounds that it was adding too much extra complexity for not enough benefit (indices and values are already "internal" APIs, which the average user will not be using.)
